Vet Clinic DB | Join Tables #4

PraisesPJMT opened the pull request:
# Vet Clinic Database - Milestone #4 | Join Tables
## Project Description
This project involves adding some many-to-many relationships and writing more complex queries to answer analytical questions.

## Feature Description
- [X] [Created a `vets` table with list of vet doctors](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/41db623ade6880e0a5ca941e96785a0fc8bac5e7)
- [X] [Created a `specialities` table with a list of vet doctors and their specializations](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/41db623ade6880e0a5ca941e96785a0fc8bac5e7)
- [X] [Created a `visits` table with a log of animals visits](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/41db623ade6880e0a5ca941e96785a0fc8bac5e7)
- [X] [Inserted four(4) data into the `vets` table](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/4cbf770ea25b38605268f76645d4644314b823b7)
- [X] [Inserted three(3) data into the `specialties` table](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/4cbf770ea25b38605268f76645d4644314b823b7)
- [X] [Inserted twenty(20) data into the `visits` table](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/4cbf770ea25b38605268f76645d4644314b823b7)
- [Wrote query codes to answer the following questions](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/commit/1eeee0eac2e7e1f98421df070b28cc65fc421b8c)
  - [X] **Query to know who was the last animal seen by `William Tatcher`**
![Screenshot from 2022-10-20 11-32-24](https://user-images.githubusercontent.com/101870993/196931366-4d508a40-4685-4318-bd90-87912e39f00d.png)

  - [X] **Query to know how many different animals `Stephenie Mendez` see**
![Screenshot from 2022-10-20 11-32-49](https://user-images.githubusercontent.com/101870993/196931382-02028163-12a7-4110-bb08-c61f43d5779d.png)

  - [X] **Query to list all vets and their specialties, including vets with no specialties**
![Screenshot from 2022-10-20 11-33-13](https://user-images.githubusercontent.com/101870993/196931411-89d2c5fe-189b-43cb-8675-2988a3ba60b4.png)

  - [X] **Query to list all animals that visited `Stephenie Mendez` between April 1st and August 30th, 2020**
![Screenshot from 2022-10-20 11-33-42](https://user-images.githubusercontent.com/101870993/196931426-60c7db3b-97c0-4b90-8291-92ddcd08d03b.png)

  - [X] **Query to know what animal has the most visits to vets**
![Screenshot from 2022-10-20 11-34-15](https://user-images.githubusercontent.com/101870993/196931451-6ee85872-c03e-42c0-957d-3ec3f23c390b.png)

  - [X] **Query to know who was Maisy Smith's first visit**
![Screenshot from 2022-10-20 11-34-40](https://user-images.githubusercontent.com/101870993/196931474-a1d6ccb2-61c5-4f6f-8d3d-b100610a82d0.png)

  - [X] **Query to show details for the most recent visit animal information, vet information, and date of visited**
![Screenshot from 2022-10-20 11-35-12](https://user-images.githubusercontent.com/101870993/196931495-e630df40-d9f5-4eb8-a933-1a3ab81189c9.png)

  - [X] **Query to know how many visits a vet that did not specialize in that animal's species**
![Screenshot from 2022-10-20 11-35-35](https://user-images.githubusercontent.com/101870993/196931525-80a536ab-c399-4f49-93ce-2bf88f502f82.png)

  - [X] **Query to know what speciality should `Maisy Smith` consider getting: Based on species she gets the most.**
![Screenshot from 2022-10-20 11-35-58](https://user-images.githubusercontent.com/101870993/196931548-b2c23154-73b0-45fe-9bde-467d49ae1d10.png)


Happy Code Reviewing :+1: :slightly_smiling_face:

PraisesPJMT committed:
Update schema.sql with codes for addition of 3 more table
Includes codes for;
- Creating 'vets' table in 'vet_clinic' database
- Creating 'specialties' table in 'vet_clinic' database
- Creating 'visits' table in 'vet_clinic' database

PraisesPJMT committed:
Update data.sql file with code for adding data into vets, visits and specialties tables
Includes codes for
- Inserting four(4) data into the 'vets' table, Each data having a name, age, and date_of_graduation
- Inserting three(3) data into the 'specialties' table, Each data having a specie, and vet
- Inserting twenty(20) data into the 'visits' table, Each data having a animal, vet, and date_visited

PraisesPJMT committed:
Update queries.sql file with more code for analysis
Includes code for;
- Query to know who was the last animal seen by `William Tatcher`
- Query to know how many different animals did `Stephenie Mendez` see
- Query to list all vets and their specialties, including vets with no specialties
- Query to list all animals that visited `Stephenie Mendez` between April 1st and August 30th, 2020
- Query to know what animal has the most visits to vets
- Query to know who was Maisy Smith's first visite
- Query to show details for the most recent visit animal information, vet information, and date of visited
- Query to know how many visits were a vet that did not specialize in that animals species
- Query to know what specialty should `Maisy Smith` consider getting: Based on species she gets the most.

sja-thedude commented on schema.sql:
```diff
@@ -57,4 +57,36 @@ ALTER TABLE animals ADD COLUMN species_id INT references species(id);
+*/
+CREATE TABLE vets (
+    id  BIGSERIAL  NOT NULL,
+    name   TEXT NOT NULL,
```
- [x] Kindly make sure that name is of varchar datatype.

PraisesPJMT replied:
OKAY! I will do just that!

sja-thedude commented on schema.sql:
```diff
@@ -57,4 +57,36 @@ ALTER TABLE animals ADD COLUMN species_id INT references species(id);
+
+/* Creating 'specialties' table in 'vet_clinic' database
+   Table has columns: id, specie, and vet
+*/
+CREATE TABLE specialties (
+    id  BIGSERIAL  NOT NULL,
+    vet TEXT,
+    specie TEXT,
+    PRIMARY KEY(id)
+);
```
- [x] Kindly make sure that name of the table is specializations.
- [x] Add only species_id & vet_id columns inside the table which is of int data type.

PraisesPJMT replied:
OKAY! I just changed the name of the table to `specialization`. Thanks :slightly_smiling_face:

PraisesPJMT replied:
Hello @sja-thedude! The instruction on the project details does not specify what data type to use for the relation of the `vet` and `specie`.
![Screenshot from 2022-10-20 12-49-01](https://user-images.githubusercontent.com/101870993/196940669-a26baddc-f278-4a2a-82b5-cfa8a17edb78.png)
I choose to use text-based type with names (`vet` & `specie`) because it is easier to read than having numbers in form of `id`.

Please can you explain to me why you think using the numbered ID is better?

sja-thedude replied:
- [x] I have advised to use it as its a join table and ids are taken to access it. And ids take int as data type.

PraisesPJMT replied:
OK! Thanks so much for the clearification.

sja-thedude commented on schema.sql:
```diff
@@ -57,4 +57,36 @@ ALTER TABLE animals ADD COLUMN species_id INT references species(id);
+   Table has columns: id, animal, vet, and date_of_visited
+*/
+CREATE TABLE visits (
+    id  BIGSERIAL  NOT NULL,
+    animal TEXT,
+    vet TEXT,
```
- [x] Kinldy make sure to Remove id, animal and vet columns and add animal_id & vet_id columns inside the table which is of int data type.

PraisesPJMT replied:
OKAY! I just removed the ID! Thanks.

PraisesPJMT replied:
Hello @sja-thedude! Same here! The instruction on the project details does not specify what data type to use for the relation of the `animal` and `vet`.
![Screenshot from 2022-10-20 12-54-09](https://user-images.githubusercontent.com/101870993/196941578-55760990-be2b-41de-9759-1d04351846cd.png)

I choose to use text-based type with names (`animal` & `vet`) because it is easier to read than having numbers in form of `id`.

Please can you explain to me why it is necessary to use IDs instead of text?

sja-thedude replied:
- [x]  I have advised to use it as its a join table and ids are taken to access it. And ids take int as data type.

PraisesPJMT replied:
I just removed the ID column for the `visits` table.
I initially used an ID to serve as a number log for registering all visits to the clinic.
Please help me share your opinion about using the ID in this table. But I have already taken it OFF.

Here is my table with the ID:
![Screenshot from 2022-10-20 13-22-19](https://user-images.githubusercontent.com/101870993/196949193-85131bfc-ce7c-44db-89a4-e8aee4bb89d5.png)

Here is the table without the ID:
![Screenshot from 2022-10-20 13-23-26](https://user-images.githubusercontent.com/101870993/196949207-7782ae09-0169-4c8c-b6be-f20bdbb2d0ff.png)

sja-thedude reviewed (changes requested):
## Required Changes ♻️

Hi @PraisesPJMT, This is @sja-thedude your code reviewer.

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

![768070f3-bc9a-4aba-b8bc-25793e00ea1e_text](https://user-images.githubusercontent.com/70269507/159483175-2f61461d-b352-4cb4-98f8-b173e05ab5ff.gif)

## Highlights
✔️ Professional Readme
✔️ Correct Gitflow

## Required 
- [ ] _Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me [@sja-thedude] in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

PraisesPJMT committed:
Change the name of the `specialties` table to `specializations`

PraisesPJMT committed:
Changed data type of `string` columns to be of variable character type.

PraisesPJMT committed:
Chnage data for `specializations` and `visits` table
- Creating `specializations` table in `vet_clinic` database,  Table now has columns: `vet_id`, and `species_id`
- Creating `visits` table in `vet_clinic` database, Table now has columns: `animal_id`, `vet_id`, and `date_visited`

PraisesPJMT committed:
Change data entries for `specializations` and `visits` tables
- Use Integer data type for `vet_id`, and `species_id` entries into `specializations` table
- Use Integer data type for `animal_id`, and `vet_id` entries into `visits` table

PraisesPJMT committed:
Modify queries to reflect change in data types of `specializations` and `visits` tables

PraisesPJMT commented:
# Changes From Recommendations
Thanks so much @sja-thedude for the amazing review. I finished implementing all the recommended changes.

## Feature Description
- [X] [Change the name of the specialties table to specializations](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/pull/4/commits/b001f209a7cf9e7bda238b732d2b2bbbe56662d1)
- [X] [Changed data type of string columns to be of variable character type.](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/pull/4/commits/9672f8f63df70b3acfab65ae0209dce66d07a485)
- [X] [Chnage data for specializations and visits table](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/pull/4/commits/9bf49546eb71e19febf52aea421a42a0a4d860b4)
- [X] [Modify queries to reflect the changes in data types of specializations and `visits` tables](https://github.com/PraisesPJMT/Veterinary-Clinic-DB/pull/4/commits/6069dedfce113755d6a89d8400357fafb54a4580) 
6069ded

## Query Results After Recommended Changes
- [X] **Query to know who was the last animal seen by `William Tatcher`**
![Screenshot from 2022-10-20 15-25-40](https://user-images.githubusercontent.com/101870993/196981108-f27f3785-332b-4069-a400-f1ec853b8c99.png)

- [X] **Query to know how many different animals `Stephenie Mendez` see**
![Screenshot from 2022-10-20 15-26-05](https://user-images.githubusercontent.com/101870993/196981128-3ebe3d88-6682-4cde-b042-1be01cc2fec5.png)

- [X] **Query to list all vets and their specializations, including vets with no specializations**
![Screenshot from 2022-10-20 15-26-29](https://user-images.githubusercontent.com/101870993/196981164-42843bc7-bd9f-4769-b490-1f1794801de2.png)

- [X] **Query to list all animals that visited `Stephenie Mendez` between April 1st and August 30th, 2020**
![Screenshot from 2022-10-20 15-26-59](https://user-images.githubusercontent.com/101870993/196981187-d17dfde4-4279-48b7-8163-4a30cb6203ba.png)

- [X] **Query to know what animal has the most visits to vets**
![Screenshot from 2022-10-20 15-27-30](https://user-images.githubusercontent.com/101870993/196981209-5d644d8d-3bc2-46ef-9e0d-c8dc3073b82a.png)

- [X] **Query to know who was Maisy Smith's first visit**
![Screenshot from 2022-10-20 15-28-07](https://user-images.githubusercontent.com/101870993/196981229-e58e3934-881d-462b-a213-a5db4a30a8d0.png)

- [X] **Query to show details for the most recent visit animal information, vet information, and date of visited**
![Screenshot from 2022-10-20 15-28-31](https://user-images.githubusercontent.com/101870993/196981263-9d1fa25d-f414-4433-95bd-50fa21800fd1.png)

- [X] **Query to know how many visits a vet that did not specialize in that animal's species**
![Screenshot from 2022-10-20 15-30-13](https://user-images.githubusercontent.com/101870993/196981287-4abd23f5-4968-4340-8953-16bb95d9c66d.png)

- [X] **Query to know what speciality should `Maisy Smith` consider getting: Based on species she gets the most.**
![Screenshot from 2022-10-20 15-30-34](https://user-images.githubusercontent.com/101870993/196981317-8498a38e-f003-408b-b768-5a3acd80b43b.png)


Happy Code Reviewing :+1: :slightly_smiling_face:

m05e5 reviewed (approved):
Hi @PraisesPJMT , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
![goodjob](https://user-images.githubusercontent.com/60631456/130260474-3c2d5e5c-96df-43de-a0db-8d0059c10862.gif)

## Highlights
✔️ your linters are all ok
✔️ Descriptive PR 

## Optional suggestions


_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

_**m05e5**_

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

PraisesPJMT commented:
Thanks so much, @m05e5 for the review!
